### PR TITLE
contrib/mysql: add classic protocol support

### PR DESCRIPTION
**Pull request by pablogonzalezpe**
## Summary

This PR adds a new `scapy.contrib.mysql` module implementing support for the MySQL classic protocol over TCP.

The current scope covers a first usable subset of the protocol for dissection and packet building, including:

- packet framing (`payload_length` + `sequence_id`)
- `Protocol::HandshakeV10`
- `Protocol::SSLRequest`
- `Protocol::HandshakeResponse41`
- `OldAuthSwitchRequest`
- `AuthSwitchRequest`
- `AuthSwitchResponse`
- `AuthMoreData`
- `OK_Packet`
- `ERR_Packet`
- `EOF_Packet`
- `COM_QUERY`
- `COM_STMT_PREPARE_OK`
- packet definitions for text resultset metadata and text rows for explicit use

Regression tests are added in `test/contrib/mysql.uts`.

## Scope

This PR is intentionally limited to a usable MVP for the classic protocol.

Automatic server-side stream inference for resultset metadata and text rows is intentionally left out of this PR for now.

Not implemented in this first version:
- TLS-encrypted MySQL payloads after `SSLRequest`
- compression
- binary resultsets
- advanced server-side stream inference for resultset metadata and rows
- full command coverage
- full authentication plugin coverage
- full prepared statement execution support

### UTScapy

Added regression tests for:
- handshake packets
- auth switch and auth more data packets
- OK / ERR / EOF packets
- `COM_QUERY`
- explicit resultset packet definitions
- text rows with multiple columns and `NULL`
- stream framing with deprecated EOF / OK terminators
- `COM_STMT_PREPARE_OK`
- multi-packet streams
- dispatcher fallbacks and stream completeness

Result:
- `UTScapy`: `19/19` passing

### Lint / typing

Validated locally with:
- `tox -e flake8` on Windows: OK
- `tox -e mypy` on Windows: OK
- `UTScapy` on WSL Ubuntu / Python 3.8: OK

### Real PCAP validation

The contrib was also tested against these public MySQL captures:

1. `umitproject/packet-manipulator/audits/pcap-tests/mysql.pcap`
2. `colinnewell/pcap2mysql-log/test/captures/big-data.pcap`
3. `arkime/tests/pcap/mysql-allow.pcap`

In local validation, common handshake/auth/query/OK/ERR/EOF packets from these captures were successfully parsed. Advanced server-side resultset metadata / row inference is intentionally left out of this PR.

## Notes

This PR follows the earlier discussion in #4954.

I tried to keep names close to the MySQL protocol naming where possible, while avoiding collisions with Scapy internals when needed.

Feedback is welcome on:
- whether this is the right MVP boundary for `scapy.contrib`
- whether any naming should be adjusted further
- whether the omitted server-side resultset inference should come in a follow-up PR

### CODE REVIEW

File: scapy/contrib/mysql.py [hunk 1]
```diff
@@ -0,0 +1,965 @@
...
+
+    def getfield(self, pkt: Packet, s: bytes) -> Tuple[bytes, Any]:
+        value, size = _read_lenenc_int(s)
+        return s[size:], value
+
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return repr(val)
+
+
+class MySQLCapabilityFlagsField(LEIntField):
```
Comment (polybassa) [inline scapy/contrib/mysql.py:300]: Is this "redefinition" of a FlagsField really necessary?

File: scapy/contrib/mysql.py [hunk 2]
```diff
@@ -0,0 +1,965 @@
...
+    def __init__(self, name: str, default: int = 0) -> None:
+        LEShortEnumField.__init__(self, name, default, MYSQL_CHARACTER_SETS)
+
+
+class MySQLColumnFlagsField(LEShortField):
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return _flag_repr(int(val), MYSQL_COLUMN_FLAGS)
+
+
+class MySQLLenEncStrField(Field[Any, Any]):
```
Comment (polybassa) [inline scapy/contrib/mysql.py:325]: Whats the purpose of this field?

File: scapy/contrib/mysql.py [hunk 3]
```diff
@@ -0,0 +1,965 @@
...
+
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return repr(val)
+
+
+class MySQLAuthResponseField(Field[Any, Any]):
+    """
+    Authentication response encoding depends on client capabilities.
+
+    - CLIENT_PLUGIN_AUTH_LENENC_CLIENT_DATA: string<lenenc>
```
Comment (polybassa) [inline scapy/contrib/mysql.py]: It's not the scapy way to introduce a new Field for such a handling. More correct would be to use "ConditionalFields" or MultipleTypeFields

File: scapy/contrib/mysql.py [hunk 4]
```diff
@@ -0,0 +1,965 @@
...
+        end = s.find(b"\x00")
+        if end < 0:
+            return b"", s
+        return s[end + 1:], s[:end]
+
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return repr(val)
+
+
+class MySQLTextRowValuesField(Field[Any, Any]):
```
Comment (polybassa) [inline scapy/contrib/mysql.py:367]: Is this Field really necessary?

**Comment by polybassa** [comment]
Please try to increase the test coverage, since this looks like AI-generated code.

**Comment by pablogonzalezpe** [comment]
Thanks, that is fair feedback.

I will add more tests to exercise the currently uncovered branches in \scapy.contrib.mysql\. The current version already includes UTScapy regression tests and validation against real MySQL pcaps, but I agree that the patch coverage can be improved further.

**Comment by pablogonzalezpe** [comment]
I added more UTScapy tests to exercise uncovered branches, including auth-response variants, helper edge cases, fallback/error paths, and incomplete stream reassembly.

File: scapy/contrib/mysql.py (hunk 3 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py]: That is fair, thanks. I refactored this part to use `MultipleTypeField` in `MySQLHandshakeResponse41` instead of a dedicated `MySQLAuthResponseField`.

The auth response still supports the same three encodings (`lenenc`, `1-byte length`, and NUL-terminated), but the field selection is now expressed in a way that is closer to Scapy conventions.

I also re-ran the MySQL UTScapy tests as well as `flake8` and `mypy` after the change.

File: scapy/contrib/mysql.py (hunk 1 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py:300]: The goal here was not to redefine `FlagsField` semantics, but to keep this serialized as a regular little-endian integer while providing a readable representation in `show()`.

I used this small wrapper because MySQL capability flags are 32-bit little-endian values and the current code keeps the raw integer behavior unchanged. If you prefer using a native `FlagsField` pattern here, I can adjust it.

File: scapy/contrib/mysql.py (hunk 4 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py:367]: This field is used for text resultset rows, which are encoded as a variable-length sequence of length-encoded values, with `0xfb` representing `NULL`.

The goal was to keep parsing and building symmetric for row values without introducing a separate packet per column. If you prefer another Scapy pattern for this, I can revisit it.

File: scapy/contrib/mysql.py (hunk 2 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py:325]: This field is used for MySQL `string<lenenc>` values, where the string length is encoded using MySQL's length-encoded integer format.

I kept it as a small field because this encoding is reused in several MySQL structures, such as column definitions and connection attributes, and I did not find an existing Scapy field that directly matched this format. If there is a preferred existing Scapy pattern for this, I can adjust it.

File: scapy/contrib/mysql.py (hunk 1 above)
Comment (polybassa) [reply scapy/contrib/mysql.py]: A standard FlagsField can provide LE representation:

```
class FlagsField(_BitField[Optional[Union[int, FlagValue]]]):
    """ Handle Flag type field
   ...
   ...

   :param name: field's name
   :param default: default value for the field
   :param size: number of bits in the field (in bits). if negative, LE                   <-----
   :param names: (list or str or dict) label for each flag
       If it's a str or a list, the least Significant Bit tag's name
       is written first.

File: scapy/contrib/mysql.py [hunk 5]
```diff
@@ -0,0 +1,960 @@
...
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return repr(val)
+
+
+class MySQLCapabilityFlagsField(LEIntField):
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return _flag_repr(int(val), MYSQL_CLIENT_FLAGS)
+
+
+class MySQLStatusFlagsField(LEShortField):
```
Comment (polybassa) [inline scapy/contrib/mysql.py]: same here, please reuse existing fields.

File: scapy/contrib/mysql.py [hunk 6]
```diff
@@ -0,0 +1,960 @@
...
+    0x0400: "TIMESTAMP",
+    0x0800: "SET",
+}
+
+
+def _capability(flags: int, mask: int) -> bool:
+    return bool(flags & mask)
+
+
+def _flag_repr(value: int, mapping: Any) -> str:
```
Comment (polybassa) [inline scapy/contrib/mysql.py]: This function is already part of FlagsField. 

remove it

File: scapy/contrib/mysql.py [hunk 7]
```diff
@@ -0,0 +1,960 @@
...
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return _flag_repr(int(val), MYSQL_CLIENT_FLAGS)
+
+
+class MySQLStatusFlagsField(LEShortField):
+    def i2repr(self, pkt: Optional[Packet], val: Any) -> str:
+        return _flag_repr(int(val), MYSQL_STATUS_FLAGS)
+
+
+class MySQLCharsetField(ByteEnumField):
```
Comment (polybassa) [inline scapy/contrib/mysql.py]: Please use existing Field definitions, if the derived class is not providing "new" functionality.

File: scapy/contrib/mysql.py [hunk 8]
```diff
@@ -0,0 +1,960 @@
...
+    def __init__(self, name: str, default: int = 0) -> None:
+        ByteEnumField.__init__(self, name, default, MYSQL_CHARACTER_SETS)
+
+
+class MySQLShortCharsetField(LEShortEnumField):
+    def __init__(self, name: str, default: int = 0) -> None:
+        LEShortEnumField.__init__(self, name, default, MYSQL_CHARACTER_SETS)
+
+
+class MySQLColumnFlagsField(LEShortField):
```
Comment (polybassa) [inline scapy/contrib/mysql.py]: Use FlagsField.

File: scapy/contrib/mysql.py (hunk 1 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py]: Thanks for pointing this out.

I pushed a refactor that replaces the custom flag/enum wrapper fields with standard Scapy fields:
- `FlagsField(..., -32, MYSQL_CLIENT_FLAGS)`
- `FlagsField(..., -16, MYSQL_STATUS_FLAGS)`
- `FlagsField(..., -16, MYSQL_COLUMN_FLAGS)`
- `ByteEnumField` / `LEShortEnumField` for character sets

I also removed the custom `_flag_repr` helper and adjusted `_capability()` so it works with `FlagValue` instances returned by `FlagsField`.

I re-ran the MySQL UTScapy tests, `flake8`, `mypy`, and `git diff --check` locally after the change.

**Comment by polybassa** [comment]
Please fix the unit test.

**Comment by pablogonzalezpe** [comment]
Thanks, fixed.

I reproduced the failure with \conf.debug_dissector=True\. The fallback test was using a short EOF-like server payload (\e00\), which could raise while being parsed as an incomplete EOF packet in strict dissection mode.

I changed the test to use an unambiguous unknown server payload for the Raw fallback case and re-ran the MySQL UTScapy tests locally, including the \-K scanner\ run and strict dissection mode.

**Comment by polybassa** [comment]
@gpotter2 What do you think?

File: scapy/contrib/mysql.py [hunk 9]
```diff
@@ -0,0 +1,936 @@
...
+    def guess_payload_class(self, payload: bytes) -> type:
+        return MySQLTextResultSetRow
+
+
+class _MySQLServerEOFPacket(MySQLServerPacket):
+    def guess_payload_class(self, payload: bytes) -> type:
+        return MySQLEOFPacket
+
+
+def _mysql_client_cls(
```
Comment (gpotter2) [inline scapy/contrib/mysql.py]: There are too many functions that are useless or only used once. This one typically shouldn't exist

File: scapy/contrib/mysql.py [hunk 10]
```diff
@@ -0,0 +1,936 @@
...
+        if header == 0x01 and payload_length > 1:
+            return MySQLServerPacket
+        if _can_parse_column_definition(payload):
+            return _MySQLServerColumnDefinitionPacket
+        if header != 0xFB:
+            return _MySQLServerResultSetColumnCountPacket
+    return MySQLServerPacket
+
+
+def _mysql_stream_complete(data: bytes) -> bool:
```
Comment (gpotter2) [inline scapy/contrib/mysql.py]: I would include this in tcp_reassemble since it's only used there

File: scapy/contrib/mysql.py [hunk 11]
```diff
@@ -0,0 +1,936 @@
...
+def _mysql_server_cls(
+    pkt: Packet,
+    lst: Any,
+    cur: bytes,
+    remain: bytes,
+) -> Optional[type]:
+    if len(remain) < 4:
+        return None
+    payload_length = struct.unpack("<I", remain[:3] + b"\x00")[0]
+    payload = remain[4:4 + payload_length]
```
Comment (gpotter2) [inline scapy/contrib/mysql.py]: I am a bit confused by this code. It calls 3 or 4 functions once that all iterate through a list of stuff. Couldn't this be merged in a single iteration?

File: scapy/contrib/mysql.py [hunk 12]
```diff
@@ -0,0 +1,936 @@
...
+    state_items = list(lst)
+    if cur is not None:
+        state_items.append(cur)
+    field_list_state = _mysql_server_field_list_state(state_items)
+    if field_list_state is not None:
+        if payload[:1] == b"\xFE":
+            return _MySQLServerEOFPacket
+        if _can_parse_column_definition(payload):
+            return _MySQLServerColumnDefinitionPacket
```
Comment (gpotter2) [inline scapy/contrib/mysql.py]: I'm not sure I understand what those classes are for. Like `_MySQLServerColumnDefinitionPacket` ? Why not use the one underneath?

File: scapy/contrib/mysql.py (hunk 4 above)
Comment (gpotter2) [reply scapy/contrib/mysql.py]: It looks very much like a PacketListField. You should try to make one by extending `i2m` instead of rewriting it entirely.

**Review by gpotter2 — COMMENTED** [review]
It's a good start but the code needs a cleaning up phase. It has way too many functions that are used only once and sometimes not that useful, which looks very AI-ish. It probably can be fixed easily though.

**Comment by pablogonzalezpe** [comment]
Thanks for the feedback. I cleaned up the module structure to reduce one-off helpers and make the stream dispatch/reassembly logic more direct and easier to follow, without changing the supported protocol scope.

I re-ran the MySQL UTScapy tests, strict dissection mode, flake8, mypy, the WSL CI-like runs, and the real-PCAP smoke tests used during development. Everything is still passing.

**Comment by gpotter2** [comment]
Thanks for the update. There are still comments you haven't addressed :)

File: scapy/contrib/mysql.py (hunk 4 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py]: Thanks, I reworked this part to use a more native Scapy pattern.

`MySQLTextRowValuesField` now reuses `FieldListField`, with a small scalar field handling a single text-row value (`lenenc` string or `0xfb` for NULL), instead of implementing the whole list parsing/building logic in one custom field.

I re-ran the MySQL UTScapy tests, strict dissection mode, flake8, mypy, and the real-PCAP smoke tests we used during development. Everything is still passing.

File: scapy/contrib/mysql.py (hunk 10 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py]: Done, I moved that logic into `tcp_reassemble()` so it is no longer kept as a separate helper.

File: scapy/contrib/mysql.py (hunk 12 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py]: I removed those wrapper packet classes and now keep the forced server-side payload selection in the stream reassembly/dispatch path instead.

File: scapy/contrib/mysql.py (hunk 11 above)
Comment (pablogonzalezpe) [reply scapy/contrib/mysql.py]: I simplified that part as well and collapsed the server-side state handling into a more direct single-pass dispatch flow.

**Comment by pablogonzalezpe** [comment]
Hi, just a small follow-up on this PR.

I believe the review comments were addressed in the latest updates, but I am happy to make further adjustments if needed.

**Comment by polybassa** [comment]
Perfect, thanks!

File: scapy/contrib/mysql.py [hunk 13]
```diff
@@ -0,0 +1,885 @@
...
+    remain: bytes,
+) -> Optional[type]:
+    if len(remain) < 4:
+        return None
+    return MySQLClientPacket
+
+
+# Track enough server-side state to force metadata/resultset packet types
+# during TCP stream reassembly.
+def _mysql_server_cls(
```
Comment (polybassa) [inline scapy/contrib/mysql.py]: Could you please rework this code or move it to a dedicated PR. 

All above parts of the PR look fine and would be ready to merge.

**Comment by pablogonzalezpe** [comment]
I reduced the scope of the server-side stream inference in this PR to keep it focused on the mergeable MVP discussed in review. The advanced heuristics for automatically inferring resultset metadata and text rows from server streams are now out of scope for this PR. The module still provides framing, handshake/auth support, query packets, common OK/ERR/EOF responses, and the packet definitions for explicit use. I also updated the PR description and reran UTScapy, flake8, and mypy after the change. If it makes sense, I can keep the omitted server-side resultset inference for a separate follow-up PR.

**Comment by polybassa** [comment]
Thanks! Feel free to open a new PR for the additional features.

**Review by polybassa — APPROVED** [review]
